src/MultipartStreamBuilder.php

```diff
@@ -133,28 +131,44 @@ public function addResource($name, $resource, array $options = [])
      */
     public function build()
     {
-        $streams = '';
+        // Open a temporary read-write stream as buffer.
+        // If the size is less than predefined limit, things will stay in memory.
+        // If the size is more than that, things will be stored in temp file.
+        $buffer = fopen('php://temp', 'r+');
         foreach ($this->data as $data) {
             // Add start and headers
-            $streams .= "--{$this->getBoundary()}\r\n".
-                $this->getHeaders($data['headers'])."\r\n";
+            fwrite($buffer, "--{$this->getBoundary()}\r\n".
+                $this->getHeaders($data['headers'])."\r\n");
 
-            // Convert the stream to string
-            /* @var $contentStream StreamInterface */
+            /**
+             * @var \Psr\Http\Message\StreamInterface
+             */
             $contentStream = $data['contents'];
+
+            // Read stream into buffer
             if ($contentStream->isSeekable()) {
-                $streams .= $contentStream->__toString();
+                $contentStream->rewind(); // rewind to beginning.
+            }
+            if ($contentStream->isReadable()) {
+                while (!$contentStream->eof()) {
+                    // read 1MB chunk into buffer until reached EOF.
+                    fwrite($buffer, $contentStream->read(1048576));
+                }
             } else {
-                $streams .= $contentStream->getContents();
+                // Try to getContents for non-readable stream.
```

Comment:
this doesn't read well

Reply:
Trying to mimic the behaviour of [Laminas](https://github.com/laminas/laminas-httphandlerrunner/blob/4c952795f1ee141aa4c1d817995d6f76015ea07d/src/Emitter/SapiStreamEmitter.php#L61-L77) (previously Zend framework) here.

Reply:
Or if you're referring to the comment line, please advice how to rewrite it.

Reply:
I was referring to the line I commented on. The comment text.